Alt Asset Preload and Persistence (aka "Error 404: Asset Load Hiccup Not Found")

Malkierian opened the pull request:
This adds alt asset preloading to the flow, as well as some common-sense alt asset persistence (menus, effects, audio, skyboxes). This altogether eliminates the asset load hitch that alt asset mods commonly suffer from except in a very few specific circumstances.

This all now runs in the background, whether alt assets are enabled or not, with no adverse effects on performance aside from the continued higher memory usage.

This does make it load a lot more assets in the persistent manner, so when using the 4K pack, usage can be 8GB+ on a continual basis, but I'm sure the 4K users would prefer to have that with virtually no hiccups rather than lower memory usage. Obviously, the usage will be significantly lower with standard HD packs.

This has the following effects (demonstrated with OoT Reloaded 4K to properly showcase the capabilities): audio on!

### Startup is significantly faster when loading with alt assets enabled, and the logo, transitions, and sound effects all load without issues.

#### Before:

https://github.com/user-attachments/assets/557e17f9-2441-4926-9e20-2e5ccf7d7c32

#### After:

https://github.com/user-attachments/assets/4562ee24-e4b5-463a-a78f-90f0e5e1f0d6

### Loading directly into a save file is also still significantly smoother:

#### Before:

https://github.com/user-attachments/assets/84c7c7cc-6d93-4a35-98b5-06fd33546914

#### After:

https://github.com/user-attachments/assets/d2e587f3-1730-40c5-96e0-62949d181ad0

### Loading into file select is faster and smoother, and there are no skybox load hiccups

#### Before:

https://github.com/user-attachments/assets/ab04fd91-bae1-4236-8fa6-c50e05e8f0f7

#### After:

https://github.com/user-attachments/assets/cf6e1daa-83ec-4da1-81ac-ed46e8b7c010

### The pause menu, and visual effects, give no hitches on first load, even across transitions

#### Before:

https://github.com/user-attachments/assets/ac1fc302-d686-458e-8a11-84fed589466d

#### After:

https://github.com/user-attachments/assets/74dfc333-d097-4751-87eb-f17883cff65e

### Reset loads are blazing fast

#### Before:

https://github.com/user-attachments/assets/64ac049e-90d7-4113-a7b3-d538df165391

#### After:

https://github.com/user-attachments/assets/1b83a442-1fef-4bbc-a990-c2ca7e2e6b7e

### And finally, the transition hitches are virtually eliminated, only being noticeable when loading areas like Hyrule Field, and toggling alt assets on in new areas takes almost no time at all:

#### Before:

https://github.com/user-attachments/assets/84b5bac3-28d8-470a-a9fc-2701730ace73

#### After:

https://github.com/user-attachments/assets/5f5108c9-e35b-4d4d-80cf-14cd42d17076

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2611961835.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2611965955.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2611974068.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Introduces targeted scene alt asset loading and unloading, as well as some persistent alt asset loading. Utilizes threading to handle LUS directory load and unload filtering.

Malkierian committed:
A few more persistent preload order changes.

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -11,9 +11,18 @@
 
+bool delayedAltAssetLoad = false;
+s16 unloadScene = -1;
+std::shared_ptr<BS::thread_pool> rmhThreadPool = std::make_shared<BS::thread_pool>(1);
```
what does `rmh` mean? resource manager hooks?

i know this is a bit of a nit but this name feels less than ideal to me. someone unfamiliar would have to guess what `rmh` means, but the variable name spells out `ThreadPool` (which can be inferred from the type)

i don't have a perfect suggestion for a name but something like `resourceThreads` might work? i'd love to hear any ideas others have

Malkierian replied:
I'm down for changing it. It was named for the fie it's in, ResourceManagerHelpers.

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -395,49 +404,49 @@ extern "C" char* ResourceMgr_LoadArrayByNameAsVec3s(const char* path) {
 
-        return (char*)data;
+    return (char*)data;
     // }
```
i know commenting out the caching stuff isn't part of this PR, but this feels like a good chance to just remove the commented out code

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -395,49 +404,49 @@ extern "C" char* ResourceMgr_LoadArrayByNameAsVec3s(const char* path) {
 
 extern "C" CollisionHeader* ResourceMgr_LoadColByName(const char* path) {
-    return (CollisionHeader*) ResourceGetDataByName(path);
+    return (CollisionHeader*)ResourceGetDataByName(path);
```
seeing all of these whitespace changes just reminds me of how we need to get clang-format into ship

Malkierian replied:
LOL. This only happened because I cut and pasted a couple of times to get IntelliSense to fix itself XD

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -503,3 +512,139 @@ extern "C" void ResourceMgr_ClearSkeletons() {
 }
+
+bool IsSharedScene(int16_t sceneNum) {
+    return sceneNum > SCENE_INSIDE_GANONS_CASTLE || sceneNum == SCENE_GANONS_TOWER || sceneNum == SCENE_THIEVES_HIDEOUT;
```
i think being explicit here could be nice. listing out all the scenes that aren't shared makes it easier to tell from a glance what "shared" means, as opposed to leading someone unfamiliar to go digging into `scene_table.h` to understand what this is checking for.

it's a bit more verbose (11 lines) this way but it doesn't feel bad imo

```suggestion
    return sceneNum != SCENE_DEKU_TREE &&
           sceneNum != SCENE_DODONGOS_CAVERN &&
           sceneNum != SCENE_JABU_JABU &&
           sceneNum != SCENE_FOREST_TEMPLE &&
           sceneNum != SCENE_FIRE_TEMPLE &&
           sceneNum != SCENE_WATER_TEMPLE &&
           sceneNum != SCENE_SPIRIT_TEMPLE &&
           sceneNum != SCENE_SHADOW_TEMPLE &&
           sceneNum != SCENE_BOTTOM_OF_THE_WELL &&
           sceneNum != SCENE_ICE_CAVERN &&
           sceneNum != SCENE_GERUDO_TRAINING_GROUND;
```

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -503,3 +512,139 @@ extern "C" void ResourceMgr_ClearSkeletons() {
+std::string GetScenePath(int16_t sceneNum) {
+    std::string sceneName = gSceneTable[sceneNum].sceneFile.fileName;
+    std::string path = "alt/scenes/shared/" + sceneName + "/*";
+    if (!IsSharedScene(sceneNum)) {
+        size_t pos = 0;
+        if ((pos = path.find("/shared/", 0)) != std::string::npos) {
```
just curious here, why do we need to check the path for `/shared/` when we've already checked `!IsSharedScene`

Malkierian replied:
It's replacing shared. The original starts with shared in it, this finds the position of it to insert mq or nonmq. I don't know why it checks for npos there, I copied it from your WIP XD

briaguya0 replied:
i wish i could ask my past self why this was in there, because current me definitely doesn't know

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -503,3 +512,139 @@ extern "C" void ResourceMgr_ClearSkeletons() {
+    ResourceLoadDirectoryAsync("alt/code/*");
+}
+
+extern "C" void ResourceMgr_LoadPersistentAltAssets() {
```
so is the idea here that whatever someone resets back to will never be unloaded?

Malkierian replied:
Nothing in the startup gets unloaded, except maybe spot00, I just don't know. The skyboxes take up the most memory by far, and that's the main common thing between pre-game and in-game. Everything else is mostly incidental to each other.

Malkierian replied:
Oh, I get what you're asking. To clarify what's going on here, the Fast File Select options determine the initial load priority for some of the persistent assets, but they all get persisted eventually with the Delayed load.

briaguya0 replied:
gotcha, that makes sense. everything spinning logo/title screen/file select related never unloads, but this just makes sure they preload appropriately

tangentially related: it'd probably be nice to add a summary of *what* never unloads to the PR description

briaguya0 commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -11,9 +11,18 @@
 #include <Fast3D/gfx_pc.h>
 #include <DisplayList.h>
 
+bool delayedAltAssetLoad = false;
```
it looks like this is always false?

Malkierian replied:
Ah, yeah, I forgot I'd removed that path because I was able to thread the delayed persistent assets after the block in LoadPersistenAssets.

briaguya0 reviewed (commented):
haven't done any playtesting but overall the code and comparison videos look great! left a few comments/questions/suggestions

Malkierian committed:
Address review comments, add rando MQ handling.

briaguya0 reviewed (approved):
i still haven't playtested this but all my code comments have been addressed!

Malkierian committed:
Fix alt asset toggle not saving the CVar.

Malkierian committed:
Fix scene unload registration (no HF hiccups!).
Add comment documentation.
Refine a little persistent preload.

Malkierian commented:
Fixed an issue that was causing the game to unload the scene it was going into if it was a back and forth (otherwise just lagging behind by a scene), but I'm marking this as do not merge as I want to try to refine the persistence to cut down on memory usage.

Malkierian committed:
Adapt to LUS multi-filter functions.
Un-persist skybox textures, add dynamic skybox loading instead.

Malkierian committed:
Bump LUS, add documentation, remove magic time of day numbers, add dynamic skybox loading and unloading.

Malkierian commented:
Now requires <https://github.com/Kenix3/libultraship/pull/752> before I can remove do not merge, but dynamic skybox loading and unloading is done. Reloaded 4K is now buttery smooth, and for me never got above 6GB of RAM usage thanks to the skybox tweaks.

Malkierian commented:
Of course, ymmv, since the preload time might not be enough on lower-end machines, but should still improve things quite a bit.

Malkierian committed:
Rework skybox load/unload timing to prevent load/unload chains.
Update for related LUS PR changes, which includes removing the threading (will be re-implemented when LUS Context thread pool access is added).

Malkierian committed:
Bump to latest LUS.

Malkierian committed:
Re-treading the threading.

Malkierian committed:
Swap `Load/UnloadResources` calls to non-async within threaded helper functions to prevent debug stutter on transition.

Malkierian committed:
Final LUS bump.

Archez commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1438,6 +1443,8 @@ extern "C" void Graph_ProcessGfxCommands(Gfx* commands) {
     bool curAltAssets = CVarGetInteger(CVAR_ENHANCEMENT("AltAssets"), 0);
     if (prevAltAssets != curAltAssets) {
         prevAltAssets = curAltAssets;
+        CVarSetInteger(CVAR_ENHANCEMENT("AltAssets"), prevAltAssets);
```
This CVarSet is doing nothing useful. `prevAltAssets` is equal to `curAltAssets` at this point, and `curAltAssets` was retrieved via a CVarGet.

Archez commented on soh/soh/z_play_otr.cpp:
```diff
@@ -76,6 +77,9 @@ void OTRPlay_InitScene(PlayState* play, s32 spawn) {
     Object_InitBank(play, &play->objectCtx);
     LightContext_Init(play, &play->lightCtx);
     TransitionActor_InitContext(&play->state, &play->transiActorCtx);
+    //ResourceMgr_LoadAllSceneResources(play->sceneNum, false);
```
This line not needed anymore?

Malkierian replied:
Nope, it is not. That's triggered at the start of transition now. XD

Archez commented on soh/soh/z_play_otr.cpp:
```diff
@@ -76,6 +77,9 @@ void OTRPlay_InitScene(PlayState* play, s32 spawn) {
     LightContext_Init(play, &play->lightCtx);
     TransitionActor_InitContext(&play->state, &play->transiActorCtx);
+    //ResourceMgr_LoadAllSceneResources(play->sceneNum, false);
+    ResourceMgr_UnloadSceneAssets();
```
Can this just live in the OnSceneInit hook? Since this just submits a task for the resource thread, then execution time doesn't seem important.

Malkierian replied:
That's a fair point, it probably can.

Archez commented on soh/src/code/game.c:
```diff
@@ -467,11 +467,22 @@ void GameState_Destroy(GameState* gameState) {
-        ResourceUnloadDirectory("alt/*");
-        gfx_texture_cache_clear();
+        // Performing clear skeletons before unload resources fixes an actor heap corruption crash due to the skeleton patching system.
+        ResourceMgr_ClearSkeletons();
```
Before we always fired this when destroying the game state, now you have it only if the scene has changed during destroy. This probably should moved up to always execute like before.

Archez commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -503,3 +506,279 @@ extern "C" void ResourceMgr_ClearSkeletons() {
+        // Title screen music
+        Ship::Context::GetInstance()->GetResourceManager()->LoadResourceAsync("audio/sequences/030_Title_Theme");
+        Ship::Context::GetInstance()->GetResourceManager()->LoadResourceAsync("audio/fonts/06_Title_Theme");
+    }
+    else if (skipTitle && fastFile == 4) {
```
Just a nit pick, but since you did a lot of white spacing fixes in this file, it would be great to also fix these `} else` lines to be on the same line.

Archez replied:
There are still a bunch of
```c
}
else if {
```
that should be set onto the same line (specifically new code introduced by this PR for this file).

Malkierian replied:
Oh, how'd I miss those...

Malkierian committed:
Address review.

Archez commented on soh/src/code/game.c:
```diff
@@ -466,11 +466,21 @@ void GameState_Destroy(GameState* gameState) {
 
     osSyncPrintf("game デストラクタ終了\n"); // "game destructor end"
 
+    // Performing clear skeletons before unload resources fixes an actor heap corruption crash due to the skeleton patching system.
```
This comment was duplicated

Archez commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1438,6 +1443,7 @@ extern "C" void Graph_ProcessGfxCommands(Gfx* commands) {
     bool curAltAssets = CVarGetInteger(CVAR_ENHANCEMENT("AltAssets"), 0);
     if (prevAltAssets != curAltAssets) {
         prevAltAssets = curAltAssets;
+        Ship::Context::GetInstance()->GetWindow()->GetGui()->SaveConsoleVariablesNextFrame();
```
When I asked for the CVarSet to be removed, I also meant for this save cvar on next frame to be removed too.

Malkierian replied:
Good point.

Malkierian committed:
Address review redux.

Malkierian committed:
manual bump back to latest LUS

aMannus commented:
I believe I asked this recently as well but I don't remember the answer. Is this ready to go? Is it dependant on anything else? Is there open issues with it?

Malkierian commented:
No issues that I know of, but still waiting on @Archez having a deeper look, and he had more pressing issues crop up in the last week or so.

Malkierian committed:
Add vanilla bulk asset preload.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
Fix ResourceManagerHelpers initialization order.
Rename `LoadDelayedPersistentAltAssets` to `LoadDelayedPersistentAssets`.
Remove audio path from LoadDelayedPersistentAssets, as that's already handled in OTRAudio_Init.
Exclude textures/buttons from LoadDelayedPersistentAssets (PNGs, don't have a factory).

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
Implement ResourceManager and ResourceHelper thread pool shut down sequences.

Malkierian committed:
LUS bump

Malkierian committed:
LUS unbump

Malkierian committed:
LUS re-bump

Malkierian committed:
Redo LUS

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
Adapt to title screen changes.
Force non-async loads for startup assets.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
LUS bump.

Malkierian committed:
Make all vanilla assets load prior to triggering any alt asset loads.
Add Link to title-specific alt asset loading, and modify Hyrule Field preload to use the scene loader function.

Malkierian committed:
png exclusion mk iii

Malkierian commented:
Closing this. Irredeemable Mac issues that I just have no patience to figure out by proxy.

Malkierian committed:
Fix async load call for vanilla assets.

Malkierian committed:
force build with a comment

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into alt-asset-bulk-preload

Malkierian committed:
Make vanilla preload exact to prevent loading alt versions if alt assets is toggled.

Malkierian commented:
Removing this from the 9.0.0 milestone because it will need some form of streaming textures for Mac to not throw a fit trying to draw alt assets that haven't loaded yet, even with the preload, and I won't be able to get that done before we want to make the next release.

Malkierian commented:
Going to close this again for the foreseeable future because of the streaming texture requirement, and I just don't know when I'll get back to this.
